### PR TITLE
ODC-5577:Topology chart view automation

### PR DESCRIPTION
**Pull request by sanketpathak**
**Fix**: https://issues.redhat.com/browse/ODC-6006

**Problem**: Automation of empty state of topology chart view, topology chart view with workloads and filters in topology chart view

**Solution**: This script will automate the topology chart view page.

**Test Setup**:

Cluster should be running on local

Check the TAGS in frontend/packages/topology/integration-tests/cypress.json file be: "@topology and (@smoke or @regression) and not (@manual or @to-do or @un-verified)",

Command to execute:

yarn run dev
yarn run test-cypress-topology
Execute the topology-chart-area-visual.feature file

**Checks for approving Epic scenarios Automation PR:**

- [x] Execute the @to-do tagged gherkin scripts manually
- [x] Convert the @to-do gherkin scripts to cypress automation scripts
- [x] Once scripts are automated, remove tag @to-do
- [ ] Execute the scripts in Remote cluster

**Browser**
Chrome 89

**Test Execution Screenshot:**
Screenshot from 2021-07-04 00-05-08

### CODE REVIEW

**Comment by gajanan-more** [comment]
Screenshot 2021-07-07 at 5 30 11 PM

All the tests are running fine but I am getting this error and @sanketpathak is not. Adding /lgtm for running tests successfully.

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+Then('user sees Topology page with message "No resources found"', () => {
+  cy.get(topologyPO.emptyText).should('be.visible');
+});
+
+Given('user is at the Topology page', () => {
+  navigateTo(devNavigationMenu.Topology);
+});
+
+When('user clicks on "Add page" link in the topology page', () => {
```
Comment (makambalaji) [inline frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:21]: same comment as mentioned above

**Review by makambalaji — CHANGES_REQUESTED** [review]

File: frontend/packages/topology/integration-tests/support/page-objects/topology-po.ts [hunk 2]
```diff
@@ -24,13 +25,14 @@ export const topologyPO = {
     addNewAnnotations: '[data-test="add-button"]',
     connector: '[data-test-id="edge-handler"]',
     filterByResource: {
-      filterByResourceDropDown: '.pf-c-select__toggle-text',
-      deploymentResource: '.co-m-resource-icon.co-m-resource-deployment',
-      deploymentConfigResource: '.co-m-resource-icon.co-m-resource-deploymentconfig',
```
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/page-objects/topology-po.ts:29]: They have been replaced by the function `typeOfResource`

File: frontend/packages/topology/integration-tests/support/page-objects/topology-po.ts [hunk 3]
```diff
@@ -216,3 +218,14 @@ export const topologyPO = {
     tickButton: '[data-test-id="check-icon"]',
   },
 };
+
+export function typeOfResource(type: string) {
+  return `[data-test="${type}"]`;
+}
+
```
[the remark this answers is not in the document]
Comment (makambalaji) [reply frontend/packages/topology/integration-tests/support/page-objects/topology-po.ts:225]: you are trying to return data-test attribute  using this function and passing it to cy.get() function right, if that is the case. we can directly pass it to [cy.byTestID() ](https://github.com/openshift/console/blob/master/frontend/packages/integration-tests-cypress/support/selectors.ts#L21-L23)

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import { topologyPO, typeOfResource, typeOfWorkload } from '../../page-objects/topology-po';
+import { topologyHelper } from '../../pages/topology';
+
+When('user navigates to Topology page', () => {
+  navigateTo(devNavigationMenu.Topology);
+});
+
+Then('user sees Topology page with message "No resources found"', () => {
+  cy.get(topologyPO.emptyText).should('be.visible');
+});
```
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts]: Updated

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts (hunk 1 above)
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts]: Updated

File: frontend/packages/topology/integration-tests/support/page-objects/topology-po.ts [hunk 5]
```diff
@@ -216,3 +218,14 @@ export const topologyPO = {
...
+export function typeOfWorkload(workload: string) {
+  return `[data-test~="icon"][data-test~="${workload
+    .toLowerCase()
+    .replace(' ', '')
+    .trim()}"]`;
+}
```
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/page-objects/topology-po.ts]: Updated

File: frontend/packages/topology/integration-tests/support/page-objects/topology-po.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/page-objects/topology-po.ts]: Updated

File: frontend/packages/dev-console/integration-tests/support/pages/add-flow/git-page.ts [hunk 6]
```diff
@@ -109,6 +111,9 @@ export const gitPage = {
         break;
       case 'deployment config':
       case 'Deployment Config':
+      case 'deployment-config':
+      case 'Deployment-Config':
+      case 'DeploymentConfig':
```
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/dev-console/integration-tests/support/pages/add-flow/git-page.ts:116]: It's not working

**Comment by makambalaji** [comment]
Tested on my localhost, worked fine

**Comment by makambalaji** [comment]
/approve
Please get approval from Developers

**Comment by gajanan-more** [comment]
@jerolimov  Can you please review this PR? QE Review is already done. If there is nothing, please add `approve`
Thank you.

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  navigateTo(devNavigationMenu.Add);
+  createGitWorkload(
+    'https://github.com/sclorg/nodejs-ex.git',
+    componentName,
+    'Deployment',
+    'nodejs-ex-git-app',
+  );
+});
+
+Given('user has created deployment config workload {string}', (componentName: string) => {
```
Comment (jerolimov) [inline frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:59]:
```suggestion
Given('user has created a deployment config workload {string}', (componentName: string) => {
```

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+});
+
+Given('user is at the Topology page', () => {
+  navigateTo(devNavigationMenu.Topology);
+});
+
+When('user clicks on {string} link in the empty topology page', (build: string) => {
+  cy.get(topologyPO.graph.startBuildingNewApp)
+    .should('be.visible')
+    .click();
```
Comment (jerolimov) [inline frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:41]: Little bit confusing that you don't use `build: string` here.

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    componentName,
+    'Deployment Config',
+    'nodejs-ex-git-app',
+  );
+});
+
+When('user navigates to Topology page', () => {
+  navigateTo(devNavigationMenu.Topology);
+});
+
```
Comment (jerolimov) [inline frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:72]: This is already defined above.

```suggestion
```

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+      resourceType,
+      'django-ex-git-app',
+    );
+  },
+);
+
+When('user is at Topology page chart view', () => {
+  navigateTo(devNavigationMenu.Topology);
+});
+
```
Comment (jerolimov) [inline frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:97]: Same as `When('user navigates to Topology page', ...`

And we call the both topology variants "graph" and "list". If this should ensure that this should show the graph (chart), I would expect some additional code to do that.

So can you reuse `When('user navigates to Topology page',` here as well or add an check for the topology switch button on the top right corner?

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    cy.byTestID('Deployment').click();
+  },
+);
+
+Then(
+  'user can see only the {string} workload when Deployment-Config option is selected',
+  (workload: string) => {
+    cy.byTestID('DeploymentConfig').click();
+    cy.get(typeOfWorkload(workload)).should('be.visible');
+    cy.byTestID('DeploymentConfig').click();
```
Comment (jerolimov) [inline frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:126]: Is this back and forward clicks really required?

```suggestion
    cy.get(typeOfWorkload(workload)).should('be.visible');
```

File: frontend/packages/topology/src/components/page/TopologyEmptyState.tsx [hunk 12]
```diff
@@ -21,22 +21,27 @@ const TopologyEmptyState: React.FC<TopologyEmptyStateProps> = ({ setIsQuickSearc
   return (
     <EmptyState className="odc-topology__empty-state" variant={EmptyStateVariant.full}>
       <EmptyStateIcon variant="container" component={TopologyIcon} />
-      <Title headingLevel="h3" size="lg">
+      <Title headingLevel="h3" size="lg" data-test="no-resources-found">
```
Comment (jerolimov) [inline frontend/packages/topology/src/components/page/TopologyEmptyState.tsx:24]: If possible, please move this 2 lines up to the EmptyState as the complete component define this.

```tsx
    <EmptyState
      className="odc-topology__empty-state"
      variant={EmptyStateVariant.full}
      data-test="empty-state"
    >
      <EmptyStateIcon variant="container" component={TopologyIcon} />
      <Title headingLevel="h3" size="lg">
```

File: frontend/packages/topology/src/filters/KindFilterDropdown.tsx [hunk 13]
```diff
@@ -74,7 +74,12 @@ const KindFilterDropdown: React.FC<KindFilterDropdownProps> = ({
         </Button>
       </span>
       {kindFilters.map((filter) => (
-        <SelectOption key={filter.id} value={filter.id} isChecked={filter.value}>
+        <SelectOption
+          key={filter.id}
+          value={filter.id}
+          isChecked={filter.value}
+          data-test={filter.labelKey ? t(filter.labelKey) : filter.label}
```
Comment (jerolimov) [inline frontend/packages/topology/src/filters/KindFilterDropdown.tsx:81]: I think we don't want and need to translate `data-test` attributes.

```suggestion
          data-test={filter.label}
```

File: frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx [hunk 14]
```diff
@@ -154,6 +154,7 @@ const BaseNode: React.FC<BaseNodeProps> = ({
             paddingX={8}
             paddingY={4}
             kind={kind}
+            data-test={`name-${element.getLabel()}`}
```
Comment (jerolimov) [inline frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx:157]: I don't find any code which is looking for this `name-` data-test attribute?

**Review by jerolimov — CHANGES_REQUESTED** [review]
Added some findings. Please take a look and let me know what you think.

File: frontend/packages/console-shared/src/components/quick-search/QuickSearchBar.tsx [hunk 15]
```diff
@@ -36,6 +36,7 @@ const QuickSearchBar: React.FC<QuickSearchBarProps> = ({
         onChange={onSearch}
         autoFocus={autoFocus}
         value={searchTerm}
+        data-test="quick-search-input"
```
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/console-shared/src/components/quick-search/QuickSearchBar.tsx:39]: Changed to input, It will be used for automation in future

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts (hunk 11 above)
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:126]: It's not required but I did to make it consistent with the previous then statement and if in future further test scenarios are added in the file, then we do not have to worry about the checks being disabled

File: frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx (hunk 14 above)
Comment (sanketpathak) [reply frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx:157]: I was trying out something so this attribute was required but it didn't work out. I've not removed it because it can be used for future automation

File: frontend/packages/topology/src/components/page/TopologyEmptyState.tsx [hunk 16]
```diff
@@ -21,22 +21,27 @@ const TopologyEmptyState: React.FC<TopologyEmptyStateProps> = ({ setIsQuickSearc
...
             }}
           >
             Start building your application
           </Button>{' '}
-          or visit the <Link to="/add">Add page</Link> for more details.
+          or visit the{' '}
+          <Link to="/add" data-test="add-page">
+            Add page
+          </Link>{' '}
+          for more details.
```
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/topology/src/components/page/TopologyEmptyState.tsx:44]: Updated with the first change

File: frontend/packages/topology/src/components/page/TopologyEmptyState.tsx (hunk 12 above)
Comment (sanketpathak) [reply frontend/packages/topology/src/components/page/TopologyEmptyState.tsx:24]: Updated

File: frontend/packages/topology/src/filters/KindFilterDropdown.tsx (hunk 13 above)
Comment (sanketpathak) [reply frontend/packages/topology/src/filters/KindFilterDropdown.tsx:81]: Updated

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts (hunk 10 above)
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:97]: Updated with checking the switch button

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts (hunk 9 above)
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:72]: Sorry I missed that, removed

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts (hunk 7 above)
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:59]: Updated

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts (hunk 8 above)
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:41]: Updated

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts (hunk 11 above)
Comment (jerolimov) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts:128]: Sorry don't get the edge case where it is needed to click items twice.

Also, this is an `Then(...` condition. I would suggest not to click an items here anymore. If a click before that is needed this should be part of a `When` statement, or?

I would suggest to remove all clicks from

```tsx
Then(
  'user can see only the {string} workload when Deployment option is selected',
```

and

```tsx
Then(
  'user can see only the {string} workload when Deployment-Config option is selected',
```

File: frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx (hunk 14 above)
Comment (jerolimov) [reply frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx:157]: But then I would recommend another `data-test`. Please remove it from this PR when it is not used and then you can add it with a test scenario where it is used.

**Review by jerolimov — CHANGES_REQUESTED** [review]
Looks good so far. But please address also this two comments:

1. https://github.com/openshift/console/pull/9399#discussion_r673755216
2. https://github.com/openshift/console/pull/9399#discussion_r674036702

File: frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx (hunk 14 above)
Comment (sanketpathak) [reply frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx:157]: Sure Christoph, will update it

Comment (sanketpathak) [reply frontend/packages/topology/src/components/graph-view/components/nodes/BaseNode.tsx]: Updated

File: frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts (hunk 11 above)
Comment (sanketpathak) [reply frontend/packages/topology/integration-tests/support/step-definitions/topology/chart-view.ts]: Updated

File: frontend/packages/dev-console/integration-tests/support/pages/add-flow/git-page.ts [hunk 17]
```diff
@@ -230,12 +232,14 @@ export const gitPage = {
     }
   },
   verifyValidatedMessage: (gitUrl = 'https://github.com/sclorg/nodejs-ex.git') => {
-    cy.get(gitPO.gitSection.validatedMessage).should('not.have.text', 'Validating...');
+    cy.get(gitPO.gitSection.validatedMessage)
+      .should('not.include.text', 'Validating...')
+      .and('not.include.text', 'Repository URL to build and deploy your code from source');
```
Comment (makambalaji) [inline frontend/packages/dev-console/integration-tests/support/pages/add-flow/git-page.ts:237]: Can you please add this text to [static-text ](https://github.com/openshift/console/tree/master/frontend/packages/dev-console/integration-tests/support/constants/staticText)folder

**Review by makambalaji — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/integration-tests/support/pages/add-flow/git-page.ts (hunk 17 above)
Comment (sanketpathak) [reply frontend/packages/dev-console/integration-tests/support/pages/add-flow/git-page.ts]: Added

File: frontend/packages/topology/src/filters/TopologyFilterBar.tsx [hunk 18]
```diff
@@ -92,7 +92,7 @@ const TopologyFilterBar: React.FC<TopologyFilterBarProps> = ({
           </ToolbarItem>
         </ToolbarGroup>
         <ToolbarGroup variant={ToolbarGroupVariant['filter-group']}>
-          <ToolbarItem>
+          <ToolbarItem data-test={'filter-by-resource'}>
```
Comment (jerolimov) [inline frontend/packages/topology/src/filters/TopologyFilterBar.tsx:95]:
```suggestion
          <ToolbarItem data-test="filter-by-resource">
```

**Review by jerolimov — APPROVED** [review]
@sanketpathak @makambalaji  I run all topology tests with `yarn run test-cypress-topology`, feature which is touched in this PR is green, but all others are red. Sanket says that you Balaji will check these later / in another PR.

image

The frontend changes are fine for me! Approving them. Balaji, please add lgtm if the rest is fine for you.

/approve

**Comment by makambalaji** [comment]
Executed the test scenarios as per the pr description, working as expected
image

**Comment by jerolimov** [comment]
Readd Balajis lgtm as the latest changes are requested by myself and looks fine. :+1: 

/lgtm
